zkwasm testing script

MCJOHN974 opened the pull request:
(WIP)
Here I added script for easy run and test zkasm. Currently it need proverjs installed in same folder as wasmtime is. Further I will add automatic mode that downlads all necessary files (or maybe add some scripts from proverjs to this repo) for running zkwasm and proverjs preinstalled will not needed. Also will try to add this script to CI.

nagisa reviewed (approved):
Looks neat so far!

MCJOHN974 commented:
(WIP)
I tried to add some flags that will be useful in CI and in other usecases, but for now flags --install-zkwasm and --install-zkwasm-permanent don't work properly. Anyway, you still can use script without this flags with proverjs preinstalled, it should work.

nagisa commented on cranelift/test-zk.sh:
```diff
@@ -0,0 +1,111 @@
+#!/bin/bash
+
```
One thing that I typically suggest doing is to `set -eux` and `set -o pipefail`. These provide a much better error handling and debugging opportunities in a black box environment like the CI.

nagisa commented on cranelift/test-zk.sh:
```diff
@@ -0,0 +1,111 @@
+#!/bin/bash
```
```suggestion
#!/usr/bin/env bash
```

There are systems that don't store bash at `/bin` (e.g. NixOS.)

akashin commented on .github/workflows/main.yml:
```diff
@@ -38,6 +38,16 @@ concurrency:
+    - uses: actions/checkout@v3
+      with:
+        submodules: true
+    - run: ls
```
Looks like can this be removed?

akashin commented on .github/workflows/main.yml:
```diff
@@ -38,6 +38,16 @@ concurrency:
+    steps:
+    - uses: actions/checkout@v3
+      with:
+        submodules: true
```
Technically, we don't really need submodules, as we don't build Rust code in this action.

akashin commented on .github/workflows/main.yml:
```diff
@@ -38,6 +38,16 @@ concurrency:
 jobs:
   # Check Code style quickly by running `rustfmt` over all code
+  Runzk:
+    name: Test generated zkwasm
```
Nit: "zkwasm" -> "zkAsm"

akashin commented on .github/workflows/main.yml:
```diff
@@ -38,6 +38,16 @@ concurrency:
 
 jobs:
   # Check Code style quickly by running `rustfmt` over all code
+  Runzk:
```
Maybe a bit more specific "test_zkasm"?

nagisa replied:
The comment above is also wrong now.

akashin commented on ci/test-zk.sh:
```diff
@@ -0,0 +1,89 @@
+else
+    zkasm_file="$BASE_DIR/$1"
+    # Replace all "//" with "/"
+    zkasm_file="${zkasm_file//\/\//\/}"
```
Why do we need this?

akashin commented on ci/test-zk.sh:
```diff
@@ -0,0 +1,89 @@
+        echo "Cloning zkevm-rom into ../../ directory..."
+        git clone https://github.com/0xPolygonHermez/zkevm-rom/ ../../zkevm-rom > /dev/null 2>&1
+        cd ../../zkevm-rom
+        npm install
```
Consider using https://blog.npmjs.org/post/171556855892/introducing-npm-ci-for-faster-more-reliable here

MCJOHN974 commented:
Hey, @nagisa , I changed bash script a bit and also added this script to CI. As you can see [job](https://github.com/near/wasmtime/actions/runs/6301925196/job/17108045776) runs succesfully, and, [job](https://github.com/near/wasmtime/actions/runs/6301792507/job/17107623262) fails cause of error in add_memory.zkwasm.

akashin commented on ci/test-zk.sh:
```diff
@@ -0,0 +1,89 @@
+        cd ../../zkevm-rom
+        npm install
+        ;;
+    "temporary")
```
I don't think we would ever use this mode, both in local and remote context, should we remove it?

akashin reviewed (changes requested):
Looks super useful, thanks!

akashin commented on ci/test-zk.sh:
```diff
@@ -0,0 +1,89 @@
+#!/bin/bash
+
+# Colors
+RED='\033[0;31m'
```
These are not used anymore

MCJOHN974 commented:
Hey, @akashin, @nagisa I made some cleanup and the workflow runs as expected. Also i fixed all comments except [this one](https://github.com/near/wasmtime/pull/7#discussion_r1336136868), but I think it is an enchantment and can be added later

akashin reviewed (approved):
Looks good, let's ship it!

MCJOHN974 commented:
Thanks all for your reviews!
